src/adapters/convex/products/sidechain-staking/convexSidechainStakingAdapter.ts

```diff
@@ -0,0 +1,258 @@
+import {
+  IMetadataBuilder,
+  CacheToFile,
+} from '../../../../core/decorators/cacheToFile'
+import { NotImplementedError } from '../../../../core/errors/errors'
+import { buildTrustAssetIconUrl } from '../../../../core/utils/buildIconUrl'
+import { filterMapAsync } from '../../../../core/utils/filters'
+import { getTokenMetadata } from '../../../../core/utils/getTokenMetadata'
+import {
+  GetEventsInput,
+  GetPositionsInput,
+  MovementsByBlock,
+  ProtocolPosition,
+  TokenType,
+  GetPositionsInputWithMandatoryTokenAddresses,
+  AssetType,
+  PositionType,
+  ProtocolDetails,
+  GetAprInput,
+  GetApyInput,
+  ProtocolTokenApr,
+  ProtocolTokenApy,
+} from '../../../../types/adapter'
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { CONVEX_FACTORY_ADDRESS } from '../../common/constants'
+import {
+  ConvexFactorySidechain__factory,
+  ConvexRewardFactorySidechain__factory,
+} from '../../contracts'
+import { RewardPaidEvent } from '../../contracts/ConvexRewardFactorySidechain'
+import { CONVEX_TOKEN } from '../staking/convexStakingAdapter'
+
+export class ConvexSidechainStakingAdapter
+  extends LpStakingAdapter
+  implements IMetadataBuilder
+{
+  productId = 'sidechain-staking'
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'Convex',
+      description: 'Convex pool adapter',
+      siteUrl: 'https://www.convexfinance.com/',
+      iconUrl: buildTrustAssetIconUrl(Chain.Ethereum, CONVEX_TOKEN.address),
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+      productId: this.productId,
+      assetDetails: {
+        type: AssetType.StandardErc20,
+      },
+    }
+  }
+
+  @CacheToFile({ fileKey: 'protocol-token' })
+  async buildMetadata() {
+    const convexFactory = ConvexFactorySidechain__factory.connect(
+      CONVEX_FACTORY_ADDRESS,
+      this.provider,
+    )
+
+    const length = await convexFactory.poolLength()
+
+    const metadata: LpStakingProtocolMetadata = {}
+    await Promise.all(
+      Array.from({ length: Number(length) }, async (_, i) => {
+        const convexData = await convexFactory.poolInfo(i)
+
+        const [
+          convexToken,
+          underlyingToken,
+          [extraRewardTokens, extraRewardTokenManagers],
+        ] = await Promise.all([
+          getTokenMetadata(convexData.rewards, this.chainId, this.provider), // convex staking contract is missing name, symbol, decimal
+          getTokenMetadata(convexData.lptoken, this.chainId, this.provider),
+
+          this.getExtraRewardTokensMetadata(convexData.rewards),
+        ])
+
+        metadata[getAddress(convexToken.address)] = {
+          protocolToken: {
+            ...convexToken,
+          },
+          underlyingToken,
+          rewardTokens: [],
+          extraRewardTokens,
+          extraRewardTokenManagers,
+        }
+      }),
+    )
+
+    return metadata
+  }
+
+  private async getExtraRewardTokensMetadata(
```

Comment:
This should have gone in the previous PR, but shouldn't these `extra*` methods be private or protected? If they are not going to be called from `DefiProvider` they should not be public.

Reply:
Yep they are private. I missed in one of the MRs but they are private now, thanks